### ISSUE TITLE
Test (and fix) forms with non-standard charsets

### ISSUE THREAD

**Issue by moy** [MechanicalSoup/MechanicalSoup#191]
See https://stackoverflow.com/questions/48496955/mechanicalsoup-doing-a-form-post-with-iso8859-1-charset for a user having issue with non-UTF-8 charset and non-ascii characters.

Not sure exactly what should be done here, but currently:

<pre>
MechanicalSoup$ git grep -i charset
MechanicalSoup$ cd tests
MechanicalSoup/tests$ file *.py
conftest.py:              ASCII text
setpath.py:               Python script, ASCII text executable
test_browser.py:          Python script, ASCII text executable
test_form.py:             Python script, ASCII text executable
test_stateful_browser.py: Python script, ASCII text executable
utils.py:                 Python script, ASCII text executable
</pre>

So, basically we don't test at all what happens with non-ascii characters.

**Comment by krlk89** [comment]
I will look into this issue sometime this week.

**Comment by krlk89** [comment]
So.. HTML 4 set ISO-8859-1as the default character set [[1]](https://www.w3schools.com/html/html_charset.asp) and it's possible that some websites are not using the HTML 5 standard, which sets UTF-8 as the default. UTF-8 is also the default in Python 3.
Also, a form can have a accept-charset attribute, which specifies the character encodings that are to be used for the form submission. [[2]](https://www.w3schools.com/tags/att_form_accept_charset.asp)

On the first occasion I don't think we have a way of knowing how the server handles the input and I guess the sensible solution would be to update the documentation to tell the user that if they wish to change the encoding of the input, it's possible, like this:

``form.set("q", "üõöä".encode("ISO-8859-1"))`` [[3]](https://docs.python.org/3/library/stdtypes.html#str.encode)

On the second occasion I guess it's possible to change the encoding behind the scenes, but I am not sure if it's necessary and/or a good practice. Thoughts?

Also, adding the accept-charset to the request header didn't have any effect as far as I tested and as is written in Requests documentation: Requests does not change its behavior at all based on which custom headers are specified. [[4]](http://docs.python-requests.org/en/master/user/quickstart/#custom-headers)

I also responded to the Stack Overflow problem but haven't gotten an answer so far, so it's possible that there's something else going on, but the tests I did back up what I have written.

**Comment by hemberger** [comment]
Thanks for your work on this @krlk89!

My experience with encodings is limited, but based on your report I would recommend we do the following:
* Add an FAQ that shows how to change the encoding.
* Add a test to verify that the input encoding is not clobbered by MechanicalSoup (as you said, we can't control how a server will handle it).

If you'd like to continue working on this, please feel free to; otherwise, I will get to it when I have some more free time.

**Comment by krlk89** [comment]
@hemberger Thanks for the response. I will continue working on this next week.

**Comment by krlk89** [comment]
@hemberger A follow up, I haven't had time to work on it yet, but I'll get to it this weekend.
